ext/javy/javy.rb

```diff
@@ -0,0 +1,170 @@
+require "rbconfig"
+require "open-uri"
+require "zlib"
+require "open3"
+
+class Javy
+  ROOT = __dir__
+  TARGET = File.join(ROOT, "javy")
+  VERSION = File.read(File.join(ROOT, "version")).strip
+
+  def self.install(**args)
+    new.install(**args)
+  end
+
+  def self.build(**args)
+    new.build(**args)
+  end
+
+  def install(**args)
+    Install.call(target: target, platform: platform, version: VERSION, **args)
+  end
+
+  def build(source:, dest:)
+    exec(source, "-o", dest)
+  end
+
+  private
+
+  def exec(*args, **kwargs)
+    system(target, *args, **kwargs)
+  end
+
+  def platform
+    @platform ||= Platform.new
+  end
+
+  def target
+    @target ||= platform.format_executable_path(TARGET)
+  end
+
+  class Install
+    def self.call(target:, platform:, version:, **args)
+      new.call(target: target, platform: platform, version: version, **args)
+    end
+
+    def call(target:, platform:, version:)
+      asset = Asset.new(
+        platform: platform,
+        version: version,
+        owner: "Shopify",
+        repository: "javy",
+        basename: "javy"
+      )
+
+      downloaded = asset.download(target: target)
+      raise InstallationError.asset_not_found(platform: platform, version: version, url: asset.url) unless downloaded
+
+      raise InstallationError.installation_failed unless verify_download(target)
```

Comment:
In fact, from the security point of view, there's a bit of room for improvement in both implementations. We are assuming that what we are fetching is what we uploaded, but that might not be true. GitHub provides a fingerprint that we should use client-side to ensure we got the right asset. This is in fact what happened to codecov when they had the security incident.
This can be done in a follow-up PR though.

Reply:
Good points to both! Thanks. We don't need `verify_download`, you're right that it is irrelevant since we've configured it to be executable. I removed it for now.